src/storage.py

```diff
@@ -0,0 +1,267 @@
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase, name="storage"):
+        super().__init__(charm, name)
+        self._stored.set_default(osd_data={})
+        self.charm = charm
+        self.name = name
+
+        # Attach handlers
+        self.framework.observe(
+            getattr(charm.on, "osd_devices_storage_attached"), self._on_osd_devices_attached
+        )
+        for key in ["disk", "wal", "db"]:
+            self.framework.observe(getattr(charm.on, f"{key}_storage_attached"), self._on_attached)
+
+        # OSD Detaching handlers.
+        for key in ["osd_devices", "disk", "wal", "db"]:
+            self.framework.observe(
+                getattr(charm.on, f"{key}_storage_detaching"), self._on_storage_detaching
+            )
+
+    """handlers"""
+
+    def _on_attached(self, event: StorageAttachedEvent):
+        """Updates the attached storage devices in state."""
+        if not microceph._is_ready():
+            logger.warning("MicroCeph not ready yet, deferring storage event.")
+            event.defer()
+            return
+
+        self._clean_stale_osd_data()
+
+        enroll = {
+            "disk": [],
+            "wal": [],
+            "db": [],
+        }
+
+        # filter only storage directives for wal/db and disk.
+        accepts = {"disk", "wal", "db"}
+        for storage in [
+            x for x in self.juju_storage_list() if any(x.find(id) >= 0 for id in accepts)
+        ]:
+            # split storage names of the form disk/0
+            directive = "".join(itertools.takewhile(str.isalpha, storage))
+            storage_path = self.juju_storage_get(storage_id=storage, attribute="location")
+            if not self._get_osd_num(storage_path, directive):
+                enroll[directive].append(storage_path)
+
+        # enrolls available disks with WAL/DB and save osd data.
+        with guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_with_wal_db(disk=enroll["disk"], wal=enroll["wal"], db=enroll["db"])
+            self.charm.status.set(ActiveStatus())
+
+    def _on_osd_devices_attached(self, event: StorageAttachedEvent):
+        """Event handler for storage attach event."""
+        if not microceph._is_ready():
+            logger.warning("MicroCeph not ready yet, deferring storage event.")
+            event.defer()
+            return
+
+        self._clean_stale_osd_data()
+
+        enroll = []
+        for storage in [device for device in self.juju_storage_list() if "osd-devices" in device]:
+            path = self.juju_storage_get(storage_id=storage, attribute="location")
+            if not self._get_osd_num(path, "osd-devices"):
+                enroll.append(path)
+
+        with guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_disks_in_batch(enroll)
+            self.charm.status.set(ActiveStatus())
```

Comment:
For compatibility with zaza and friends, I suggest setting a message string such as 'charm is ready' so that external libs can know that the charm is effectively ready.

Reply:
I am getting flashbacks of your ubuntu-charm rants :D. Will fix this for storage and for bootstrap etc.